### PR TITLE
More tests for its-happening

### PR DESCRIPTION
**Pull request by chris124567**
Add a test similar to TestP2P in walletd and add some extra cases mostly relating to file contracts to the validation test.

### CODE REVIEW

File: chain/p2p_test.go [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+	return nil, nil, nil
+}
+func (h handler) Checkpoint(types.ChainIndex) (types.Block, consensus.State, error) {
+	return types.Block{}, consensus.State{}, nil
+}
+func (h handler) RelayV2Header(gateway.V2BlockHeader, *gateway.Peer)         {}
+func (h handler) RelayV2BlockOutline(gateway.V2BlockOutline, *gateway.Peer)  {}
+func (h handler) RelayV2TransactionSet([]types.V2Transaction, *gateway.Peer) {}
+
+func TestP2P(t *testing.T) {
```
Comment (lukechampine) [inline chain/p2p_test.go:104]: so the bad news is that this test only adds 1% of coverage to `chain`. Not worth it imo. I know it was a pain to get it working, sorry :(

The good news is that this test *will* be nice to have if we end up moving more syncer code into `core`. So we should keep it around in the git history for that, at least.

I think the best place to focus on for now is file contract tests in `consensus` (v1 and v2). It's crucial that we test as many edge cases as possible in the new consensus code, so we should strive for 100% coverage.

File: consensus/validation_test.go [hunk 2]
```diff
@@ -473,6 +473,22 @@ func TestValidateV2Block(t *testing.T) {
...
+				r.RenterSignature = renterPrivateKey.SignHash(cs.RenewalSigHash(*r))
+				r.HostSignature = hostPrivateKey.SignHash(cs.RenewalSigHash(*r))
+				txn.FileContractResolutions[i].Resolution = r
+				break
+			case *types.V2FileContractFinalization:
+				fc := types.V2FileContract(*r)
+				fc.RenterSignature = renterPrivateKey.SignHash(cs.ContractSigHash(fc))
+				fc.HostSignature = hostPrivateKey.SignHash(cs.ContractSigHash(fc))
+
+				txn.FileContractResolutions[i].Resolution = (*types.V2FileContractFinalization)(&fc)
```
Comment (lukechampine) [inline consensus/validation_test.go:488]: since `r` is a pointer, you don't need to set `txn.FileContractResolutions[i].Resolution = r`. For the `types.V2FileContractFinalization` branch, I would do:
```go
sigHash := cs.ContractSigHash(types.V2FileContract(*r))
r.RenterSignature = renterPrivateKey.SignHash(sigHash)
r.HostSignature = hostPrivateKey.SignHash(sigHash)
```

File: types/encoding.go [hunk 3]
```diff
@@ -598,6 +598,10 @@ func (ren V2FileContractRenewal) EncodeTo(e *Encoder) {
 	ren.HostSignature.EncodeTo(e)
 }
 
+func (fcf V2FileContractFinalization) EncodeTo(e *Encoder) {
```
Comment (lukechampine) [inline types/encoding.go:601]: docstring

File: consensus/validation_test.go [hunk 4]
```diff
@@ -428,16 +470,82 @@ func TestValidateBlock(t *testing.T) {
...
+
+			if err := consensus.ValidateBlock(cs, corruptBlock, dbStore.SupplementTipBlock(corruptBlock)); err == nil {
+				t.Fatalf("accepted block with %v", test.desc)
+			}
+		}
+	}
+}
+
+func updateProofs(cau consensus.ApplyUpdate, sces []types.SiacoinElement, sfes []types.SiafundElement, fces []types.V2FileContractElement) {
+	for i := 0; i < len(sces); i++ {
```
Comment (lukechampine) [inline consensus/validation_test.go:529]: these can be `for i := range sces`

**Comment by lukechampine** [comment]
We're above 75% coverage in consensus now, nice!
